bufferserver/src/main/java/com/datatorrent/bufferserver/internal/DataList.java

```diff
@@ -405,7 +405,27 @@ public boolean resumeSuspendedClients(final int numberOfInMemBlockPermits)
 
   public boolean isMemoryBlockAvailable()
   {
-    return (storage == null) || (numberOfInMemBlockPermits.get() > 0);
+    return (numberOfInMemBlockPermits.get() > 0) || ((storage == null) && !areSubscribersBehindByMax());
```

Comment:
IMO, overloading the meaning of `isMemoryBlockAvailable()` is confusing and should be avoided. There are two distinct cases when a Publisher is allowed to switch to a new buffer and continue reading data from the network channel. Can they be handled in the Publisher.switchToNewBuffer()?